### ISSUE TITLE
Find a more predictable way to encode the discriminant

### ISSUE THREAD

**Issue by AnttiParaoanu** [bodil/smartstring#4]
From looking at the source code it seems like smartstring makes two assumptions about std::String

First assumption is that the first element in memory is the pointer to the string buffer, and not the length nor the capacity, since the Discriminant is always stored in the first byte.

A string is just a Vec<u8> and the [documentation](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) states that the order of pointer, length and capacity is "completely unspecified", so I don't see why the pointer would always occupy the first 4/8 bytes of std::String.

The second assumption (at least on little endian systems) is that the pointer will be aligned at least on a two byte boundary. I don't see why this would always be true. While most (if not all) allocators do allocate with 16 byte alignment, I don't see why the underlying u8 Vec would have any guarantees on alingment (for example with a custom allocator).

How does smallstring ensure that these assumptions are true? Or does smallstring not assume these things?

**Comment by bodil** [comment]
It does assume these things, obviously, but it has no way to validate the assumptions consistently. It's quite possible that a future version of rustc, or even a current version of rustc on certain architectures, would choose to give the `String` a different layout, which would break this crate on little-endian archs. While there's not much that can be done about that, it does suggest some extra bounds checking to ensure such breakage doesn't trigger any memory unsafety issues.

**Comment by gretingz** [comment]
I have a suggestion for an alternative layout which doesn't assume anything and should be equally fast. The idea is that instead of storing the discriminating bit in the pointer, it is instead stored in the size or capacity field. Since collections in rust cannot be longer than isize::MAX, the most significant bit will be 0 when the string is boxed, and it can be 1 when it is inlined. The most significant byte will then store the discriminating bit and the length of the inlined array.

I'd be happy to implement this if you would be interested in merging it.

**Comment by bodil** [comment]
My first reaction is that if the discriminant isn't in the first or last byte, it's going to be impossible to implement Deref<str> for it?

**Comment by gretingz** [comment]
Yeah, you're right. The idea is to not use mem::transmute and string, but instead fix the layout, and then hope that the compiler can optimize it, in case the layouts between std::String and smartstring are identical. With the fixed layout, on little endian machines the discriminant is on the last byte and on big endian it's in the first one.

**Comment by bodil** [comment]
Well, if you can implement it, I'd certainly be interested in merging it. 😊

**Comment by bodil** [comment]
Meanwhile, just to note I added a bounds check on `InlineString::len()`, to make sure the worst thing that can happen if layout assumptions break down with the current impl is an immediate panic rather than OOB memory access.

**Comment by pickfire** [comment]
Is this implementation same as facebook string in c++?

**Comment by bodil** [comment]
I couldn't say, I haven't looked at it, but it's the same general principle.

**Comment by gretingz** [comment]
Update: I have done most of the implementation. It does have a lot of changes unfortunately. 

I have also implemented null pointer optimizations, which means that an `Option<SmartString<Mode>>` is the same size as just `SmartString<Mode>`. When mode is Compact, the discriminant is stored in the size field. Then the size field can never be zero: if the string is boxed, size cannot be zero. It the string is inlined, the discriminant has at least one bit set, so the field can't be zero.

When mode is LazyCompact, the discriminant is stored in the capacity field. This might possibly maybe have theoretically a small performance impact, so there is a feature lazy_null_pointer_optimizations that, when disabled, will store the discriminant in the size field, but disables null pointer optimizations.

The new implementation is still WIP, but I've made a commit on my [branch](https://github.com/gretingz/smartstring), which contains a rough test-passing version.
